### ISSUE TITLE
Batch size on different worker could be different?

### ISSUE THREAD

**Issue by Young768** [tensorflow/tensorflow#29975]
I want to use Tensorflow for training in a distributed environment. But current Allreduce framework requires that both the tensor dimensions and batch size should the same among all the workers. I want to know if Tensorflow supports different workers have different batch size in the same batch, but their tensor size is still the same.

**Comment by Young768** [comment]
Any luck to have your answer?

**Comment by guptapriya** [comment]
Hi, what type of distributed training are you using? You should be able to use a different batch size in each worker. all reduce is usually on the gradient which doesn't really get affected by the batch dimension. what are you trying to use the all reduce for?

**Comment by Young768** [comment]
@guptapriya Yep. I mean no matter PS or allreduce, Can I change the batch size adaptively based on their capabilities? But remain their global batch size the same in each iteration. Whats the related API to do this?

**Comment by Young768** [comment]
@guptapriya Any luck to have your answer?

**Comment by guptapriya** [comment]
Do you mean you want the same global batch size each step, but split the batch unevenly across the replicas - because one of the GPUs is for e.g. more powerful than the other? If yes, then this is possible to do on your own, but not supported easily by the high level APIs.
By default, when using high level APIs like Keras, we will try to split your batch equally among the replicas. If you have your training loop etc (using strategy.experimental_run_v2), you can provide input that is skewed like you want, and it should still work.

**Comment by Young768** [comment]
@guptapriya Yep!!! Thats what I want! Do u have any examples like this?

**Comment by guptapriya** [comment]
A short example w TF2. Can this be adapted for your use case? Basically you can change mk_input to return whatever input you want to return for a given replica. Note that when using this mechanism, you will not be able to benefit from some of the tf.data performance optimizations.. 

```
import tensorflow as tf

def mk_input():
  replica_id = tf.distribute.get_replica_context().replica_id_in_sync_group
  return replica_id * 2 

def step_fn(i):
  return i

strategy = tf.distribute.MirroredStrategy(["/cpu:0", "/gpu:0"])

for _ in range(10):
  pre_replica_inputs = strategy.experimental_run_v2(mk_input)
  results = strategy.experimental_run_v2(step_fn, args=(pre_replica_inputs,))
  print(strategy.experimental_local_results(results))
```

**Comment by Young768** [comment]
Hi @guptapriya Thanks for your example. One more question: this example has two devices: cpu and gpu. From my understanding mk_input returns the desired batch size. But how can I decide which size is assigned to cpu, or gpu here? How to bind the batch size to the corresponding replica?

**Comment by guptapriya** [comment]
I just used cpu and gpu because then i was able to run it in colab easily. you can just use whatever devices you want to use (for e.g. 2 GPUs). The order in which you specify the GPUs is the replica id order. So you can use that to determine what batch size to use on which replica. mk_input can simply return the actual input. In this toy example, I just said the input will be 2 times the replica_id. But of course in reality you will use replica_id to determine the batch size. 

e.g.:
MirroredStrategy(devices=["gpu:0", "gpu:1"])
Here "gpu:0" will become your replica_id = 0, and "gpu:1" will be replica_id = 1. 

Does that help ?

**Comment by Young768** [comment]
Hi @guptapriya I am little confused by the example. I checked the [Doc](https://www.tensorflow.org/beta/tutorials/distribute/training_loops). when we use experimental_run_v2(train_step,args=(x,)), then all these two workers will consume the same batch size, x here. But for my case, I want to assign for example 32 MB to cpu and 64 MB to GPU.

**Comment by guptapriya** [comment]
The doc is assuming you're using "strategy.experimental_distribute_dataset" which does the batch splitting for you. I am recommending not using that API. I am saying create your distributed input manually with the help of experimental_run_v2 (and do not use datasets). 
in other words, if `x` is generated by `experimental_distribute_dataset`, that will be equally split. But if `x` is generated by you, you can make it whatever you want.

**Comment by Young768** [comment]
Ok. I got it. Your example has two devices. How to specify that 32 is assigned to /cpu:0 and 64 is assigned to /gpu:0?

**Comment by guptapriya** [comment]
use replica_id_in_sync_group. it's value matches the order in which you specified your devices in the mirrored strategy constructor.

**Comment by Young768** [comment]
Thanks @guptapriya So, here comes my last question, where can I pass the desired batch size to my specified devices?

**Comment by guptapriya** [comment]
Hi, sorry about the delay in responding. You need to construct the input tensors you want in this case, there is no API where you can specify the batch size. 

Maybe I can suggest something better if you show me your use case. Where is your input coming from ? Are you using a tf.data dataset?

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
